### PR TITLE
GEOMESA-1694 Cassandra refactoring to use index-api

### PR DESCRIPTION
**Pull request by elahrvivaz**

### CODE REVIEW

File: docs/user/cassandra/install.rst [hunk 1]
```diff
@@ -67,4 +67,10 @@ To install the plugins, extract the archive and copy the contents to the ``WEB-I
 directory of your GeoServer installation. You will also need to copy the JARs from the
 ``lib`` directory of your Cassandra installation into the GeoServer ``WEB-INF/lib`` directory.
 
+apache-cassandra-3.0.11.jar
+apache-cassandra-clientutil-3.0.11.jar
+apache-cassandra-thrift-3.0.11.jar
+cassandra-driver-core-3.0.0.jar
+cassandra-driver-mapper-3.0.0.jar
```
Comment (elahrvivaz) [inline docs/user/cassandra/install.rst]: i put these here as placeholders, will fix

File: build/copyright/ccri-2016.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Copyright (c) ${project.inceptionYear}-2016 ${owner}
```
Comment (jahhulbert-ccri) [inline build/copyright/ccri-2016.txt]: should we just make this 2017 ?

Comment (elahrvivaz) [inline build/copyright/ccri-2016.txt]: I had to keep 2016 in there unless we go through and update every single file to have 2017...

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/AttributeIndex.scala [hunk 3]
```diff
@@ -269,6 +269,9 @@ object AttributeIndex {
   // store 2 bytes for the index of the attribute in the sft - this allows up to 32k attributes in the sft.
   def indexToBytes(i: Int): Array[Byte] = Array((i << 8).asInstanceOf[Byte], i.asInstanceOf[Byte])
 
+  // TODO verify this
+  def bytesToIndex(b0: Byte, b1: Byte): Short = ((b0 >> 8) + b1).toShort
```
Comment (jahhulbert-ccri) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/AttributeIndex.scala]: may need to mask of negative sign of b1 if its there? hmmm ```b1 & 0xFF``` but yeah this might need a little unit test and verification.

File: geomesa-cassandra/geomesa-cassandra-tools/bin/install-cassandra-jars.sh [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#! /usr/bin/env bash
+#
+# Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Apache License, Version 2.0 which
+# accompanies this distribution and is available at
+# http://www.opensource.org/licenses/apache2.0.php.
+#
+
+if [ -z "${%%gmtools.dist.name%%_HOME}" ]; then
```
Comment (jahhulbert-ccri) [inline geomesa-cassandra/geomesa-cassandra-tools/bin/install-cassandra-jars.sh]: double brackets ```[[ ``` and ```]]```

File: docs/user/cassandra/install.rst [hunk 5]
```diff
@@ -63,8 +116,17 @@ The GeoMesa Cassandra distribution includes a GeoServer plugin for including
...
-directory of your GeoServer installation. You will also need to copy the JARs from the
-``lib`` directory of your Cassandra installation into the GeoServer ``WEB-INF/lib`` directory.
 
-Restart GeoServer after the JARs are installed.
\ No newline at end of file
+To install the plugins, extract the archive and copy the contents to the ``WEB-INF/lib``
+directory of your GeoServer installation. You will also need to install the Cassandra JARs; these
+are not bundled to allow for different versions. The distribution includes a script to download
+the JARs: ``bin/install-cassandra-jars.sh``. Call it with the path to the GeoServer ``WEB-INF/lib`` directory.
+By default It will install the following JARs:
```
Comment (mdzimmerman) [inline docs/user/cassandra/install.rst]: "It" -> "it

File: docs/user/cassandra/usage.rst [hunk 6]
```diff
@@ -16,9 +17,8 @@ DataStore such as the PostGIS DataStore or the Accumulo DataStore.  To get a con
     import java.util.Map;
 
     Map<String, ?> params = ImmutableMap.of(
-       CassandraDataStoreParams.CONTACT_POINT().getName() , "127.0.0.1:9142",
-       CassandraDataStoreParams.KEYSPACE().getName()      , "geomesa_cassandra",
-       CassandraDataStoreParams.NAMESPACE().getName()     , "mynamespace",
-       CassandraDataStoreParams.CATALOG().getName()       , "mycatalog");
+       "geomesa.cassandra.catalog.table", "127.0.0.1:9142",
+       "geomesa.cassandra.keyspace", "geomesa_cassandra",
+       "geomesa.cassandra.contact.point", "mycatalog");
```
Comment (mdzimmerman) [inline docs/user/cassandra/usage.rst]: +1 for using the key names that users would use.

File: docs/user/cassandra/geoserver.rst [hunk 7]
```diff
@@ -1,43 +1,28 @@
...
+plugin JARs into in the right directory and restart GeoServer.
 
-Once you have a Cassandra layer set up in GeoServer you can try some queries. One way to test queries against the
-GeoServer layer is to submit HTTP requests for the WMS and WFS services. For example, assuming
-you have ingested the ``example.csv`` dataset as described above and set it up as a layer in GeoServer, this request should return
-a PNG image with a single dot::
+On the "Add Store" page, select "Cassandra (GeoMesa)". The Cassandra data store requires three parameters:
 
-    http://localhost:8080/geoserver/wms?styles=&bbox=-180,-90,180,90&layers=myworkspace:example_csv&cql_filter=bbox(geom, -101, 22.0, -100.0, 24.0, 'EPSG:4326') and lastseen between 2015-05-05T00:00:00.000Z and 2015-05-10T00:00:00.000Z&version=1.3&service=WMS&width=100&request=GetMap&height=100&format=image/png&crs=EPSG:4326
+* **geomesa.cassandra.contact.point** - the connection point Cassandra, in the form ``<host>:<port>`` -
```
Comment (mdzimmerman) [inline docs/user/cassandra/geoserver.rst]: "the connection point *for* Cassandra"

File: docs/user/cassandra/contribute.rst [hunk 8]
```diff
@@ -3,13 +3,5 @@ How you can contribute
...
-  * Configurable periodicity - utilizing one-week bounds as the coarse temporal part of the row key is not optimal for
-    all data sets and workloads.  Make the coarse temporal part of the primary key configurable - i.e. day, year, etc
-  * Configurable precision in the z3 clustering key - the full resolution z3 index results in a lot of range scans.
-    We can limit the range scans by accommodating lower precision z3 indexes and pruning false positives in a
-    post-processing step.
-  * Prune false positives with push-down predicates - if we add the latitude and longitude as columns, we can prune
-    false positives by having two predicates - the first specifying the range on z3 and the second specifying the bounds on x and y.
-  * Non-point geometries - support linestrings and polygons
\ No newline at end of file
+  * Pushdown predicates - push attribute predicates down into Cassandra rather than applying them
```
Comment (mdzimmerman) [inline docs/user/cassandra/contribute.rst]: Should we even keep this section/page? We don't do this for the other data stores.

File: docs/user/cassandra/geoserver.rst [hunk 9]
```diff
@@ -1,43 +1,28 @@
...
+If you have properly installed the GeoMesa Cassandra GeoServer plugin as described
+in :ref:`install_cassandra_geoserver`, "Cassandra (GeoMesa)" should be included in the list
+under **Vector Data Sources**. If you do not see this, check that you unpacked the
+plugin JARs into in the right directory and restart GeoServer.
 
-Once you have a Cassandra layer set up in GeoServer you can try some queries. One way to test queries against the
-GeoServer layer is to submit HTTP requests for the WMS and WFS services. For example, assuming
-you have ingested the ``example.csv`` dataset as described above and set it up as a layer in GeoServer, this request should return
-a PNG image with a single dot::
+On the "Add Store" page, select "Cassandra (GeoMesa)". The Cassandra data store requires three parameters:
```
Comment (mdzimmerman) [inline docs/user/cassandra/geoserver.rst]: Can we list the other parameters that are inherited from ``GeoMesaDataStoreFactory``, even if they're not required to run the C* data store? Since they shared among multiple data stores, maybe they can go in ``/user/datastores/*`` (and maybe this should be a separate PR).

File: geomesa-cassandra/geomesa-cassandra-tools/bin/install-cassandra-jars.sh (hunk 4 above)
Comment (elahrvivaz) [inline geomesa-cassandra/geomesa-cassandra-tools/bin/install-cassandra-jars.sh]: fixed

File: docs/user/cassandra/usage.rst [hunk 10]
```diff
@@ -1,8 +1,9 @@
 Using the Cassandra DataStore Programmatically
 ==============================================
 
-Since the Cassandra DataStore is just another Geotools DataStore, you can use it exactly as you would any other Geotools
-DataStore such as the PostGIS DataStore or the Accumulo DataStore.  To get a connection to a Cassandra DataStore, use the ```DataStoreFinder```.
+Since the Cassandra DataStore is just another Geotools DataStore, you can use it exactly as
+you would any other Geotools DataStore such as the PostGIS DataStore or the Accumulo DataStore.
```
Comment (mdzimmerman) [inline docs/user/cassandra/usage.rst]: [picky] "Geotools" -> "GeoTools"

Comment (mdzimmerman) [inline docs/user/cassandra/usage.rst]: Also, maybe there's a way to rewrite this sentence so that it doesn't say "DataStore" 17 times?

File: geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/data/CassandraFeature.scala [hunk 11]
```diff
@@ -0,0 +1,18 @@
+/***********************************************************************
+* Copyright (c) 2017 IBM
```
Comment (mdzimmerman) [inline geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/data/CassandraFeature.scala]: Why is this one just copyright IBM?

File: docs/user/cassandra/install.rst (hunk 5 above)
Comment (elahrvivaz) [inline docs/user/cassandra/install.rst]: fixed

File: docs/user/cassandra/usage.rst (hunk 10 above)
Comment (elahrvivaz) [inline docs/user/cassandra/usage.rst]: replaced with wording from hbase section

File: docs/user/cassandra/geoserver.rst (hunk 9 above)
Comment (elahrvivaz) [inline docs/user/cassandra/geoserver.rst]: added - you're right we should do this for other data stores too (as a separate pr)

File: docs/user/cassandra/geoserver.rst (hunk 7 above)
Comment (elahrvivaz) [inline docs/user/cassandra/geoserver.rst]: fixed

File: docs/user/cassandra/contribute.rst (hunk 8 above)
Comment (elahrvivaz) [inline docs/user/cassandra/contribute.rst]: removed

File: geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/data/CassandraFeature.scala (hunk 11 above)
Comment (elahrvivaz) [inline geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/data/CassandraFeature.scala]: we didn't modify it :)

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/AttributeIndex.scala (hunk 3 above)
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/AttributeIndex.scala]: the `indexToBytes` method is actually broken for values > 256... I replaced them both with guava `Shorts.to/fromBytes`. Probably anyone indexing more than 256 attributes isn't working right atm.

File: docs/user/cassandra/geoserver.rst (hunk 9 above)
Comment (jahhulbert-ccri) [inline docs/user/cassandra/geoserver.rst]: ticket!

**Comment by elahrvivaz** [comment]
merged as b8309402ba195910bac19a560160bd9a97d73b81
